File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        res = []
        
        def dfs(n):
            if not n:
                return 0
            
            lh = dfs(n.left)
            rh = dfs(n.right)
            
            res.append(abs(lh-rh)<2)
            
            return max(lh,rh)+1
        
        dfs(root)
        return all(res)

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(!root)
            return true;
        bool balanced=true;
        depth(root,balanced);
        return balanced;
    }
    
    int depth(TreeNode *root, bool &bal){
        if(!root || !bal)
            return 0;
        
        int l = depth(root->left, bal);
        int r = depth(root->right, bal);
        bal &= abs(l-r)<2;
        
        return max(l,r)+1;
    }
};